File: app/src/main/java/uk/co/sundroid/activity/data/fragments/dialogs/settings/YearEventsPickerFragment.java
package uk.co.sundroid.activity.data.fragments.dialogs.settings;

import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.app.Fragment;
import android.content.Context;
import android.content.DialogInterface;
import android.content.DialogInterface.OnClickListener;
import android.content.DialogInterface.OnMultiChoiceClickListener;
import android.os.Bundle;
import uk.co.sundroid.activity.data.fragments.dialogs.OnViewPrefsChangedListener;
import uk.co.sundroid.util.prefs.SharedPrefsHelper;

public class YearEventsPickerFragment extends DialogFragment implements OnClickListener, OnMultiChoiceClickListener {

    private final boolean[] currentEvents = new boolean[8];

    public static YearEventsPickerFragment newInstance() {
        return new YearEventsPickerFragment();
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        setRetainInstance(true);// TODO Redundant?

        currentEvents[0] = SharedPrefsHelper.INSTANCE.getShowElement(getActivity(), "yearNewMoon", true);
        currentEvents[1] = SharedPrefsHelper.INSTANCE.getShowElement(getActivity(), "yearFullMoon", true);
        currentEvents[2] = SharedPrefsHelper.INSTANCE.getShowElement(getActivity(), "yearQuarterMoon", true);
        currentEvents[3] = SharedPrefsHelper.INSTANCE.getShowElement(getActivity(), "yearSolstice", true);
        currentEvents[4] = SharedPrefsHelper.INSTANCE.getShowElement(getActivity(), "yearEquinox", true);
        currentEvents[5] = SharedPrefsHelper.INSTANCE.getShowElement(getActivity(), "yearLunarEclipse", true);
        currentEvents[6] = SharedPrefsHelper.INSTANCE.getShowElement(getActivity(), "yearSolarEclipse", true);
        currentEvents[7] = SharedPrefsHelper.INSTANCE.getShowElement(getActivity(), "yearEarthApsis", true);

        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setMultiChoiceItems(
                new CharSequence[]{ "New moons", "Full moons", "Quarter moons", "Solstices", "Equinoxes", "Lunar eclipses", "Solar eclipses", "Earth aphelion and perihelion" },
                currentEvents,
                this);
        builder.setTitle("Select events");
        builder.setPositiveButton("OK", this);
        builder.setNegativeButton("Cancel", this);
        return builder.create();
    }

    @Override
    public void onClick(DialogInterface dialogInterface, int id, boolean value) {
        currentEvents[id] = value;
    }

    @Override
    public void onClick(DialogInterface dialogInterface, int button) {
        if (button == DialogInterface.BUTTON_POSITIVE) {
            SharedPrefsHelper.INSTANCE.setShowElement(getActivity(), "yearNewMoon", currentEvents[0]);
            SharedPrefsHelper.INSTANCE.setShowElement(getActivity(), "yearFullMoon", currentEvents[1]);
            SharedPrefsHelper.INSTANCE.setShowElement(getActivity(), "yearQuarterMoon", currentEvents[2]);
            SharedPrefsHelper.INSTANCE.setShowElement(getActivity(), "yearSolstice", currentEvents[3]);
            SharedPrefsHelper.INSTANCE.setShowElement(getActivity(), "yearEquinox", currentEvents[4]);
            SharedPrefsHelper.INSTANCE.setShowElement(getActivity(), "yearLunarEclipse", currentEvents[5]);
            SharedPrefsHelper.INSTANCE.setShowElement(getActivity(), "yearSolarEclipse", currentEvents[6]);
            SharedPrefsHelper.INSTANCE.setShowElement(getActivity(), "yearEarthApsis", currentEvents[7]);

            Fragment parent = getTargetFragment();
            if (parent != null && parent instanceof OnViewPrefsChangedListener) {
                ((OnViewPrefsChangedListener) parent).onViewPrefsUpdated();
            }
        }
        dismiss();
    }

}

File: app/src/main/kotlin/uk/co/sundroid/activity/data/fragments/dialogs/settings/YearEventsPickerFragment.kt
package uk.co.sundroid.activity.data.fragments.dialogs.settings

import android.app.AlertDialog
import android.app.Dialog
import android.app.DialogFragment
import android.content.Context
import android.os.Bundle
import uk.co.sundroid.activity.data.fragments.dialogs.OnViewPrefsChangedListener
import uk.co.sundroid.util.prefs.Prefs

class YearEventsPickerFragment : DialogFragment() {

    private enum class Setting(val ref: String, val displayName: String) {
        NEW_MOON("yearNewMoon", "New moons"),
        FULL_MOON("yearFullMoon", "Full moons"),
        QUARTER_MOON("yearQuarterMoon", "Quarter moons"),
        SOLSTICE("yearSolstice", "Solstices"),
        EQUINOX("yearEquinox", "Equinoxes"),
        LUNAR_ECLIPSE("yearLunarEclipse", "Lunar eclipses"),
        SOLAR_ECLIPSE("yearSolarEclipse", "Solar eclipses"),
        EARTH_APSIS("yearEarthApsis", "Earth aphelion and perihelion")
    }

    private val currentEvents = BooleanArray(Setting.values().size, { true })

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        restore(arguments)
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        Companion.save(currentEvents, outState)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        restore(savedInstanceState)

        val builder = AlertDialog.Builder(activity)
        builder.setMultiChoiceItems(
                Setting.values().map { it.displayName }.toTypedArray(),
                currentEvents,
                { _, i, value -> currentEvents[i] = value })
        builder.setTitle("Select events")
        builder.setPositiveButton("OK", { _, _ -> run {
            Setting.values().forEachIndexed { i, setting -> Prefs.setShowElement(activity, setting.ref, currentEvents[i]) }
            val parent = targetFragment
            if (parent is OnViewPrefsChangedListener) {
                (parent as OnViewPrefsChangedListener).onViewPrefsUpdated()
            }
        }})
        builder.setNegativeButton("Cancel", null)
        return builder.create()
    }

    private fun restore(bundle: Bundle?) {
        if (bundle != null) {
            val events = bundle.getBooleanArray("events")
            events?.forEachIndexed { i, on -> currentEvents[i] = on }
        }
    }

    companion object {
        fun newInstance(context: Context): YearEventsPickerFragment {
            val currentEvents = BooleanArray(Setting.values().size)
            Setting.values().forEachIndexed { i, setting -> currentEvents[i] = Prefs.showElement(context, setting.ref, true) }
            return YearEventsPickerFragment().apply {
                arguments = save(currentEvents)
            }
        }

        private fun save(events: BooleanArray, bundle: Bundle? = Bundle()): Bundle? {
            return bundle?.apply {
                putBooleanArray("events", events)
            }
        }
    }

}